### PR TITLE
support reverse proxy authentication and impersonation

### PR DESCRIPTION
**Pull request by mlbiam**
** Describe the change **

Added support for reverse proxy authentication and impersonation, where a reverse proxy can inject an `Authorization` header that is either a user's identifying token or a service account token with impersonation headers.  To support this update, changed the main mechanism for getting credentials from a `string` that represents a token to the `api.AuthInfo` struct so the additional data can be passed in.  Also updated the caching mechanism to take impersonation data when getting a connection for the user.

** Issue reference **

https://github.com/kiali/kiali/issues/3207

** Backwards incompatible? **

No, backwards compatible

** Documentation **

* Does your contribution contain user-visible changes like e.g. new or changed configuration settings?
If so, please also update README.adoc

### CODE REVIEW

File: business/dashboards.go
Comment (jmazzitelli) [inline business/dashboards.go:16]: following the kiali coding standards, third party libs go in a second group of imports between the standard go libs and the kiali libs with blank lines separating. So move this so it looks something like this:

```
...
	"sync"

	"k8s.io/client-go/tools/clientcmd/api"

	"github.com/kiali/kiali/config"
...
```

File: business/grafana.go
Comment (jmazzitelli) [inline business/grafana.go:18]: minor change - again just order the imports for consistency across the project

File: business/layer.go
Comment (jmazzitelli) [inline business/layer.go:12]: minor change - again just order the imports for consistency across the project

File: business/proxy_status.go
Comment (jmazzitelli) [inline business/proxy_status.go:7]: minor change - again just order the imports for consistency across the project

File: business/tokenreview.go
Comment (jmazzitelli) [inline business/tokenreview.go:6]: for consistency - the kiali coding standard wants the imports like this:
```suggestion
import (
	"k8s.io/client-go/tools/clientcmd/api"
	
	"github.com/kiali/kiali/kubernetes"
)
```

File: handlers/authentication.go
Comment (jmazzitelli) [inline handlers/authentication.go:114]: This is just a comment: Just making sure users elsewhere of authInfo will know to check for both nil and non-nil values of ImpersonateUserExtra.  I haven't looked closely, but just wanted to point out this value may be nil so code that reads this value may need to check for nil rather than, say, a length of 0.

File: handlers/authentication.go
Comment (jmazzitelli) [inline handlers/authentication.go:286]: Is this comment correct? Is it a "ServiceAccount" token or a token from the header? And where was it checked for validity (I guess my question is what does it mean to be valid? Does "valid" mean it can be used to successfully make requests with? Or "valid" as in "it is in there and it isn't nil"?

File: status/discover.go
Comment (jmazzitelli) [inline status/discover.go:11]: minor change - again just order the imports for consistency across the project

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
some minor things and just some questions

File: handlers/authentication.go
Comment (mlbiam) [reply handlers/authentication.go]: this comment should (and has) been updated.  I started by duplicating the `performTokenAuthentication` function and this comment was a left-over of that.  Updated the comment.

File: handlers/authentication.go
Comment (mlbiam) [reply handlers/authentication.go:114]: i checked, this particular code was just for lazy initialization, so to say "if we haven't initialed, create the map"

**Comment by jmazzitelli** [comment]
question: does this still support "openid" auth strategy? It looks like this changes the way the "dex" option installs Dex. We need this to be backward compatible and still support "openid" auth strategy (because this is what we use when we need to test with that strategy and what our CI nightly builds use when running molecule tests).

**Comment by mlbiam** [comment]
@jmazzitelli yes, nothing should change with openid auth strategy.  The only change I made to dex was to support an additional redirect url to the reverse proxy.  that will keep dex from rejecting the request from the oauth2 proxy that's deployed.  100% backwards compatible.

**Comment by jmazzitelli** [comment]
@mlbiam Thanks! Note: there is a small conflict - should be easy to fix. If you can rebase and fix that up, I'll try to test this sometime today.

**Comment by mlbiam** [comment]
Should be good to go

**Review by jmazzitelli — APPROVED** [review]
LGTM

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
"openid" molecule test now fails. I haven't looked deep, but we now get an error in the test:

```
    TASK [With the updated login cookie, make a request that requires authentication (should now return 200)] ***
fatal: [localhost]: FAILED! => {"changed": false, "connection": "close", "content": "<html>\r\n<head><title>502 Bad Gateway</title></head>\r\n<body>\r\n<center><h1>502 Bad Gateway</h1></center>\r\n<hr><center>openresty/1.15.8.2</center>\r\n</body>\r\n</html>\r\n", "content_length": "163", "content_type": "text/html", "date": "Wed, 13 Jan 2021 17:04:48 GMT", "elapsed": 0, "msg": "Status code was 502 and not [200]: HTTP Error 502: Bad Gateway", "redirected": false, "server": "openresty/1.15.8.2", "status": 502, "url": "https://192.168.99.100/kiali/api/namespaces"}
```

Got that error when running the test this way (assuming you started k8s-minikube.sh with Dex/OIDC installed as explained earlier):

```
make -e CLUSTER_TYPE=minikube -e MOLECULE_SCENARIO="openid-test" -e MOLECULE_USE_DEV_IMAGES=true -e HELM_CHARTS_REPO_PULL=false molecule-test
```

**Comment by jmazzitelli** [comment]
All other molecule tests pass.  Only thing left is to figure out why openid test is now failing. I'm seeing if some config just needs to change in the test or if it is something more sinister.

```
              accessible-namespaces-test... success [7m 34s]
     affinity-tolerations-resources-test... success [6m 19s]
                                api-test... success [4m 37s]
                      config-values-test... success [8m 14s]
                  default-namespace-test... success [3m 22s]
                            metrics-test... success [6m 20s]
                     null-cr-values-test... success [3m 39s]
                only-view-only-mode-test... success [5m 19s]
                             openid-test... FAILURE [3m 19s]
                   os-console-links-test... skipped
                              roles-test... success [9m 40s]
                    rolling-restart-test... success [5m 11s]
                              token-test... success [3m 24s]
```

File: kubernetes/client_factory.go
Comment (jmazzitelli) [inline kubernetes/client_factory.go:186]: molecule test "openid-test" has exposed a nil-pointer reference crash here. Logs:
```
            "2021/01/13 18:44:48 http: panic serving 172.17.0.2:35626: runtime error: invalid memory address or nil pointer dereference",
            "goroutine 71 [running]:",
            "net/http.(*conn).serve.func1(0xc000510320)",
            "\t/home/jmazzite/bin/go-install/go/src/net/http/server.go:1800 +0x139",
            "panic(0x140f000, 0x2237940)",
            "\t/home/jmazzite/bin/go-install/go/src/runtime/panic.go:975 +0x3e3",
            "github.com/kiali/kiali/kubernetes.getTokenHash(0x0, 0xc0003436d0, 0x40c366)",
            "\t/home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/kubernetes/client_factory.go:186 +0x37",
            "github.com/kiali/kiali/kubernetes.(*clientFactory).getClientEntry(0xc000432b40, 0x0, 0x5c2cae, 0x1398620, 0x17dd3f0)",
            "\t/home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/kubernetes/client_factory.go:144 +0x40",
            "github.com/kiali/kiali/kubernetes.(*clientFactory).GetClient(0xc000432b40, 0x0, 0x129b8af, 0xc00003bb60, 0x1398620, 0x17dd3e0)",
            "\t/home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/kubernetes/client_factory.go:135 +0x35",
            "github.com/kiali/kiali/business.Get(0x0, 0x0, 0x0, 0x0)",
            "\t/home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/business/layer.go:85 +0x5f",
            "github.com/kiali/kiali/handlers.getBusiness(0xc00059fb00, 0x78, 0x80, 0xc0002df500)",
            "\t/home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/handlers/utils.go:86 +0x65",
            "github.com/kiali/kiali/handlers.NamespaceList(0x182ed00, 0xc000146540, 0xc00059fb00)",
            "\t/home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/handlers/namespaces.go:14 +0x32",
            "net/http.HandlerFunc.ServeHTTP(0x1664b78, 0x182ed00, 0xc000146540, 0xc00059fb00)",
            "\t/home/jmazzite/bin/go-install/go/src/net/http/server.go:2041 +0x44",
            "github.com/kiali/kiali/handlers.AuthenticationHandler.Handle.func1(0x182ed00, 0xc000146540, 0xc00059fa00)",
            "\t/home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/handlers/authentication.go:599 +0x1de",
            "net/http.HandlerFunc.ServeHTTP(0xc0004a8750, 0x182ed00, 0xc000146540, 0xc00059fa00)",
            "\t/home/jmazzite/bin/go-install/go/src/net/http/server.go:2041 +0x44",
            "github.com/kiali/kiali/server.plainHttpMiddleware.func1(0x182ed00, 0xc000146540, 0xc00059fa00)",
            "\t/home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/server/server.go:127 +0x6c",
            "net/http.HandlerFunc.ServeHTTP(0xc00055a080, 0x182ed00, 0xc000146540, 0xc00059fa00)",
            "\t/home/jmazzite/bin/go-install/go/src/net/http/server.go:2041 +0x44",
            "github.com/gorilla/mux.(*Router).ServeHTTP(0xc0002c7200, 0x182ed00, 0xc000146540, 0xc00059f800)",
            "\t/home/jmazzite/source/kiali/kiali/pkg/mod/github.com/gorilla/mux@v1.7.4/mux.go:210 +0xe2",
            "github.com/NYTimes/gziphandler.GzipHandlerWithOpts.func1.1(0x182ed00, 0xc000146540, 0xc00059f800)",
            "\t/home/jmazzite/source/kiali/kiali/pkg/mod/github.com/!n!y!times/gziphandler@v1.1.1/gzip.go:342 +0x2f1",
            "net/http.HandlerFunc.ServeHTTP(0xc0005780f0, 0x182ed00, 0xc000146540, 0xc00059f800)",
            "\t/home/jmazzite/bin/go-install/go/src/net/http/server.go:2041 +0x44",
            "net/http.(*ServeMux).ServeHTTP(0xc0004f5440, 0x182ed00, 0xc000146540, 0xc00059f800)",
            "\t/home/jmazzite/bin/go-install/go/src/net/http/server.go:2416 +0x1a5",
            "net/http.serverHandler.ServeHTTP(0xc000206460, 0x182ed00, 0xc000146540, 0xc00059f800)",
            "\t/home/jmazzite/bin/go-install/go/src/net/http/server.go:2836 +0xa3",
            "net/http.(*conn).serve(0xc000510320, 0x1832600, 0xc000593180)",
            "\t/home/jmazzite/bin/go-install/go/src/net/http/server.go:1924 +0x86c",
            "created by net/http.(*Server).Serve",
            "\t/home/jmazzite/bin/go-install/go/src/net/http/server.go:2962 +0x35c",
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]

File: handlers/authentication.go
Comment (jmazzitelli) [inline handlers/authentication.go:601]: build fails here with: `handlers/authentication.go:601:15: Errorf call has arguments but no formatting directives`

**Review by jmazzitelli — APPROVED** [review]
openid-test molecule test now passes.

**Comment by jmazzitelli** [comment]
Before I merge, let me see if I can quickly throw together a molecule test for the new header auth strategy.

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
I'm blocking this PR due to the following problem the new header molecule test has exposed:

https://github.com/kiali/kiali/issues/3207#issuecomment-760590537

**Review by jmazzitelli — APPROVED** [review]
I am approving. Once all the PRs are "approved" and ready to merge, I am going to run the unit tests and molecule tests one more final time and when I see they all pass I will merge everything.

**Comment by jmazzitelli** [comment]
I think this fixes the conflicts:

```diff
diff --git a/go.mod b/go.mod
index 92461b2a..f033d514 100644
--- a/go.mod
+++ b/go.mod
@@ -19,10 +19,8 @@ require (
        github.com/mitchellh/mapstructure v1.4.0
        github.com/openshift/api v0.0.0-20200221181648-8ce0047d664f
        github.com/pborman/uuid v1.2.1 // indirect
-       github.com/prometheus/client_golang v0.9.4
-       github.com/prometheus/client_model v0.2.0 // indirect
-       github.com/prometheus/common v0.4.1
-       github.com/prometheus/procfs v0.0.10 // indirect
+       github.com/prometheus/client_golang v1.9.0
+       github.com/prometheus/common v0.15.0
        github.com/rs/zerolog v1.20.0
        github.com/stretchr/testify v1.4.0
        golang.org/x/oauth2 v0.0.0-20201109201403-9fd604954f58 // indirect
diff --git a/go.sum b/go.sum
index 8143abf9..8dab17ff 100644
--- a/go.sum
+++ b/go.sum
@@ -219,6 +219,21 @@ github.com/openshift/api v0.0.0-20200221181648-8ce0047d664f h1:ATPK7UhEwglONJc8q
 github.com/openshift/api v0.0.0-20200221181648-8ce0047d664f/go.mod h1:dh9o4Fs58gpFXGSYfnVxGR9PnV53I8TW84pQaJDdGiY=
 github.com/pborman/uuid v1.2.1 h1:+ZZIw58t/ozdjRaXh/3awHfmWRbzYxJoAdNJxe/3pvw=
 github.com/pborman/uuid v1.2.1/go.mod h1:X/NO0urCmaxf9VXbdlT7C2Yzkj2IKimNn4k+gtPdI/k=
+github.com/opentracing-contrib/go-observer v0.0.0-20170622124052-a52f23424492/go.mod h1:Ngi6UdF0k5OKD5t5wlmGhe/EDKPoUM3BXZSSfIuJbis=
+github.com/opentracing/basictracer-go v1.0.0/go.mod h1:QfBfYuafItcjQuMwinw9GhYKwFXS9KnPs5lxoYwgW74=
+github.com/opentracing/opentracing-go v1.0.2/go.mod h1:UkNAQd3GIcIGf0SeVgPpRdFStlNbqXla1AfSYxPUl2o=
+github.com/opentracing/opentracing-go v1.1.0/go.mod h1:UkNAQd3GIcIGf0SeVgPpRdFStlNbqXla1AfSYxPUl2o=
+github.com/openzipkin-contrib/zipkin-go-opentracing v0.4.5/go.mod h1:/wsWhb9smxSfWAKL3wpBW7V8scJMt8N8gnaMCS9E/cA=
+github.com/openzipkin/zipkin-go v0.1.6/go.mod h1:QgAqvLzwWbR/WpD4A3cGpPtJrZXNIiJc5AZX7/PBEpw=
+github.com/openzipkin/zipkin-go v0.2.1/go.mod h1:NaW6tEwdmWMaCDZzg8sh+IBNOxHMPnhQw8ySjnjRyN4=
+github.com/openzipkin/zipkin-go v0.2.2/go.mod h1:NaW6tEwdmWMaCDZzg8sh+IBNOxHMPnhQw8ySjnjRyN4=
+github.com/pact-foundation/pact-go v1.0.4/go.mod h1:uExwJY4kCzNPcHRj+hCR/HBbOOIwwtUjcrb0b5/5kLM=
+github.com/pascaldekloe/goe v0.0.0-20180627143212-57f6aae5913c/go.mod h1:lzWF7FIEvWOWxwDKqyGYQf6ZUaNfKdP144TG7ZOy1lc=
+github.com/pborman/uuid v1.2.0/go.mod h1:X/NO0urCmaxf9VXbdlT7C2Yzkj2IKimNn4k+gtPdI/k=
+github.com/performancecopilot/speed v3.0.0+incompatible/go.mod h1:/CLtqpZ5gBg1M9iaPbIdPPGyKcA8hKdoy6hAWba7Yac=
+github.com/peterbourgon/diskv v2.0.1+incompatible/go.mod h1:uqqh8zWWbv1HBMNONnaR/tNboyR3/BZd58JJSHlUSCU=
+github.com/pierrec/lz4 v1.0.2-0.20190131084431-473cd7ce01a1/go.mod h1:3/3N9NVKO0jef7pBehbT1qWhCMrIgbYNnFAZCqQ5LRc=
+github.com/pierrec/lz4 v2.0.5+incompatible/go.mod h1:pdkljMzZIN41W+lC3N2tnIh5sFi+IEE17M5jbnwPHcY=
 github.com/pkg/errors v0.8.0/go.mod h1:bwawxfHBFNV+L2hUp1rHADufV3IMtnDRdf1r5NINEl0=
 github.com/pkg/errors v0.8.1/go.mod h1:bwawxfHBFNV+L2hUp1rHADufV3IMtnDRdf1r5NINEl0=
 github.com/pmezard/go-difflib v1.0.0 h1:4DBwDE0NGyQoBHbLQYPwSUPoCMWR5BEzIk/f1lZbAQM=
```

**Review by jmazzitelli — APPROVED** [review]
back to normal, unit tests and openid and header-auth molecule tests ran for sanity check - all pass.

**Pull request merged**
